ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,150 @@
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.topic.SourceTopicsExtractor;
+import io.confluent.ksql.util.KsqlException;
+
+import java.util.Set;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.common.acl.AclOperation;
+
+/**
+ * Checks if a {@link ServiceContext} has access to the source and target topics of transient
+ * and persistent query statements.
+ */
+public class TopicAccessValidator {
+  private final ServiceContext serviceContext;
+  private final MetaStore metaStore;
+  private final SourceTopicsExtractor sourceTopicsExtractor;
+
+  public TopicAccessValidator(final ServiceContext serviceContext, final MetaStore metaStore) {
+    this.serviceContext = serviceContext;
+    this.metaStore = metaStore;
+    this.sourceTopicsExtractor = new SourceTopicsExtractor(metaStore);
+  }
+
+  /**
+   * Checks if the {@link ServiceContext} has access to the source and target topics of transient
+   * and persistent query statements.
+   * </p>
+   * It checks for Read access to all the source topics and Write access to the target topic.
+   * </p>
+   * Other permissions checks are already validated by the {@code Injectors} which attempts to
+   * create and/or delete topics using the same {@code ServiceContext}.
+   *
+   * @param statement The statement to verify for permissions.
+   * @throws KsqlException If a topic is not authorized for access, or the topic does not exist.
+   */
+  public void validate(final Statement statement) {
+    if (statement instanceof Query) {
+      validateQueryTopicSources((Query)statement);
+    } else if (statement instanceof InsertInto) {
+      validateInsertInto((InsertInto)statement);
+    } else if (statement instanceof CreateAsSelect) {
+      validateCreateAsSelect((CreateAsSelect)statement);
+    }
+  }
+
+  private void validateQueryTopicSources(final Query query) {
+    sourceTopicsExtractor.process(query, null);
+    for (String kafkaTopic : sourceTopicsExtractor.getSourceTopics()) {
+      checkAccess(kafkaTopic, AclOperation.READ);
+    }
+  }
+
+  private void validateCreateAsSelect(final CreateAsSelect createAsSelect) {
+    /*
+     * Check topic access for CREATE STREAM/TABLE AS SELECT statements.
+     *
+     * Validates Write on the target topic if exists, and Read on the query sources topics.
+     *
+     * The Create access is validated by the TopicCreateInjector which will attempt to create
+     * the target topic using the same ServiceContext used for validation.
+     */
+
+    validateQueryTopicSources(createAsSelect.getQuery());
+
+    String kafkaTopic;
+
+    try {
+      // Get the topic from the Metastore if a different name is used
+      kafkaTopic = getSourceTopicName(createAsSelect.getName().getSuffix());
+    } catch (final KsqlException e) {
+      // Get the topic if WITH(kafka_topic='') clause is used
```

Comment:
Initially, the getSourceTopicName failed when I called this validator in other places. I didn't look into this after refactoring it. Now that I re-run the tests, I don't see this code to fail if I don't have the catch. As you said, the source topic is on the metastore because is already created.

I removed the exception.